src/dnssec/validator/base.rs

```diff
@@ -0,0 +1,746 @@
+    ) -> Result<Digest, AlgorithmError>;
+
+    /// Return the key size in bits or an error if the algorithm is not
+    /// supported.
+    fn key_size(&self) -> Result<usize, AlgorithmError>;
+}
+
+impl<Octets> DnskeyExt for Dnskey<Octets>
+where
+    Octets: AsRef<[u8]>,
+{
+    /// Calculates a digest from DNSKEY.
+    ///
+    /// See [RFC 4034, Section 5.1.4]:
+    ///
+    /// ```text
+    /// 5.1.4.  The Digest Field
+    ///   The digest is calculated by concatenating the canonical form of the
+    ///   fully qualified owner name of the DNSKEY RR with the DNSKEY RDATA,
+    ///   and then applying the digest algorithm.
+    ///
+    ///     digest = digest_algorithm( DNSKEY owner name | DNSKEY RDATA);
+    ///
+    ///      "|" denotes concatenation
+    ///
+    ///     DNSKEY RDATA = Flags | Protocol | Algorithm | Public Key.
+    /// ```
+    ///
+    /// [RFC 4034, Section 5.1.4]: https://tools.ietf.org/html/rfc4034#section-5.1.4
+    fn digest<N: ToName>(
+        &self,
+        name: &N,
+        algorithm: DigestAlg,
+    ) -> Result<Digest, AlgorithmError> {
+        let mut buf: Vec<u8> = Vec::new();
+        with_infallible(|| {
+            name.compose_canonical(&mut buf)?;
+            self.compose_canonical_rdata(&mut buf)
+        });
+
+        let mut ctx = match algorithm {
+            DigestAlg::SHA1 => DigestContext::new(DigestType::Sha1),
+            DigestAlg::SHA256 => DigestContext::new(DigestType::Sha256),
+            DigestAlg::SHA384 => DigestContext::new(DigestType::Sha384),
+            _ => {
+                return Err(AlgorithmError::Unsupported);
+            }
+        };
+
+        ctx.update(&buf);
+        Ok(ctx.finish())
+    }
+
+    /// The size of this key, in bits.
+    ///
+    /// For RSA keys, this measures the size of the public modulus.  For all
+    /// other algorithms, it is the size of the fixed-width public key.
+    fn key_size(&self) -> Result<usize, AlgorithmError> {
+        match self.algorithm() {
+            SecAlg::RSASHA1
+            | SecAlg::RSASHA1_NSEC3_SHA1
+            | SecAlg::RSASHA256
+            | SecAlg::RSASHA512 => {
+                let data = self.public_key().as_ref();
+                // The exponent length is encoded as 1 or 3 bytes.
+                let (exp_len, off) = if data[0] != 0 {
+                    (data[0] as usize, 1)
+                } else {
+                    // NOTE: Even though this is the extended encoding of the length,
+                    // a user could choose to put a length less than 256 over here.
+                    let exp_len =
+                        u16::from_be_bytes(data[1..3].try_into().unwrap());
+                    (exp_len as usize, 3)
+                };
+                let n = &data[off + exp_len..];
+                Ok(n.len() * 8 - n[0].leading_zeros() as usize)
+            }
+            SecAlg::ECDSAP256SHA256 | SecAlg::ECDSAP384SHA384 => {
+                // ECDSA public keys have two points.
```

Comment:
The original code said that ECDSA public keys have a marker byte as well and subtracted one from the len to account for this. Is that no longer the case?